DOCSP-26001 4.11.0 compat updates (#454)
mongodb/docs-node: source/whats-new.txt

* :ref:`Version 3.7 <version-3.7>`
* :ref:`Version 3.6 <version-3.6>`

.. _version-4.11:

What's New in 4.11
------------------

New features of the 4.11 {+driver-short+} release include:

- Recursive Schema Support

  - The driver now supports recursive schemas and provides type safety

>     on dot notation queries up to a depth of 9. Beyond a depth of 9,

Reviewer: Issue: Per the style guide (Numbers versus Words), for descriptions of quantities less than 10, use spelled-out numbers unless it meets one of the exception cases listed in the aforementioned article.

```suggestion
    for dot notation queries up to a depth of nine. Beyond a depth of nine,
```

Comment applies to all instances in this PR.

>     your code still compiles but is no longer type checked and falls
>     back to a type of ``any``. This depth limit is a product of

Reviewer: Avoid passive voice when possible. To ensure active voice, check whether it is clear "what" is performing the action.

```suggestion
    TypeScript compiles your code but no longer type checks it. The type falls back to type ``any``.
```

    Typescript's recursive type limitations.

    - Create an interface with a recursive schema:

      .. code-block:: js

> 

Reviewer: Suggestion: use "copyable: false" property for all the code snippets in this section since they aren't attempting to demonstrate something that the user should try using copy and paste.

         interface CircularSchema {
             name: string;
             nestedSchema: CircularSchema;
         }

    - Suppose we have a collection of type ``Collection<CircularSchema>``.

>       Below a depth of 9, type checking is enforced. The following code

Reviewer: Avoid passive voice when possible. To ensure active voice, check whether it is clear "what" is performing the action.

```suggestion
      TypeScript enforces type checking below a depth of nine. The following code
```

>       results in a compilation error because the name must be a

Reviewer: ```suggestion
      causes a TypeScript compilation error because the name must be a
```

      ``string`` type:

      .. code-block:: js

>          collection.findOne({ 'nestedSchema.nestedSchema.nestedSchema.name': 25 })

Author: The value is an int because this could should result in a compilation error. The type of this value should mismatch the type defined in the interface `CircularSchema`

    - At a depth greater than 9, the following code compiles but is not
      type checked, as it is at a depth of 11:

      .. code-block:: js

         collection.findOne({

>              'nestedSchema.nestedSchema.nestedSchema.nestedSchema.nestedSchema.nestedSchema.nestedSchema.nestedSchema.nestedSchema.nestedSchema.name': 25

Reviewer: Issue: I think it's difficult to use this example to demonstrate depth of 11.

Suggestion: think of a shorter name for the CircularSchema type if you want to show an example and have the reader be able to count 10 instances of it.

>          })

Reviewer: Issue: I was confused as to how this sub-list is organized. Per the style guide, "introduce a list [or sub-list] with context." 

Suggestion: use sub-headers and ordered lists to organize the information.

Author: Removed all lists and presented code examples under "Feature Demonstration"

* :ref:`Version 3.7 <version-3.7>`
* :ref:`Version 3.6 <version-3.6>`

.. _version-4.11:

What's New in 4.11
------------------

The 4.11 {+driver-short+} release includes **recursive schema support**

> and provides type safety for dot notation queries up to a depth of

Reviewer: Issue: 
I think "recursive schema" is jargon that we should define. I referred to our existing docs on known limitations involving recursive types (https://www.mongodb.com/docs/drivers/node/current/fundamentals/typescript/#known-limitations)
and Googled, but I wasn't able to find the term recursive schema. This made it difficult for me to understand whether the conjunction between "recursive schema support" and "provides type safety for dot notation queries..." was intended, or whether it was a definition.

Suggestion:
I think, but am not certain, this is referring to "recursive types", but also that the recursion applies to creating types that correspond to collections that are recursive in nature. It could be worthwhile reaching out to the Node.js DBX team to ensure you can provide an accurate definition.

Also, depending on whether they're related, the conjunction could be replaced by a separate sentence that explains the term or introduces the second item (whichever applies here).

Author: Followed up and confirmed that recursive schema support refers to support for recursive types as the schema used in the `Collection` class

nine. Beyond a depth of nine, your code still compiles but TypeScript no longer
type checks and values fall back to a type of ``any``. The recursive type
depth limit is a current limitation in TypeScript.

> Feature Demonstration

Reviewer: Suggestion: describe what you're showing with the subheader text: "<specific feature / support> Example (or Examples)"

~~~~~~~~~~~~~~~~~~~~~

Suppose we have a collection of type ``Collection<CircularSchema>``:

.. code-block:: js
   :copyable: false

   interface CircularSchema {
       name: string;
       ref: CircularSchema;
   }

TypeScript enforces type checking below a depth of nine. The following
code causes a TypeScript compilation error because the ``name`` property
value must be a ``string`` type:

.. code-block:: js
   :copyable: false

   collection.findOne({ 'ref.ref.ref.name': 25 })

> At a depth greater than nine, TypeScript compiles your code but no
> longer type checks it. The type falls back to type ``any``. The

Reviewer: I think this version might be a little more clear and avoid the passive voice in the second sentence.

```suggestion
At a depth greater than nine, TypeScript compiles your code, but 
falls back to accepting the ``any`` type at that level. The
```

Author: I think the property type falls back to `any`, so I reworded to reflect that

* :ref:`Version 3.7 <version-3.7>`
* :ref:`Version 3.6 <version-3.6>`

.. _version-4.11:

What's New in 4.11
------------------

The 4.11 {+driver-short+} release includes **recursive schema support**,
meaning that you can use recursive types as the schema in the
``Collection`` class. The driver also provides type safety for dot

> notation queries up to a depth of nine in this release. Beyond a depth
> of nine, TypeScript compiles your code, but property types fall back
> to ``any`` type at that level. The recursive type depth limit is a

Reviewer: Suggestion: present the main subject first in a long phrase.

```suggestion
notation queries up to a depth of nine in this release. Typescript compiles
your code beyond a depth of nine, but falls back to the property type ``any``
at that level. The recursive type depth limit is a
```

current limitation in TypeScript.

Recursive Schema Type Checking Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Suppose we have a collection of type ``Collection<CircularSchema>``:

.. code-block:: js
   :copyable: false

   interface CircularSchema {
       name: string;
       ref: CircularSchema;
   }

TypeScript enforces type checking below a depth of nine. The following
code causes a TypeScript compilation error because the ``name`` property
value must be a ``string`` type:

.. code-block:: js
   :copyable: false

   collection.findOne({ 'ref.ref.ref.name': 25 })

At a depth greater than nine, TypeScript compiles your code but no
longer type checks it. The following code references a property at a

> depth of 11 and does not cause a compilation error:

Reviewer: Suggestion: I think an explanation of what doesn't cause the compilation error could be helpful rather than inferring from the code example. E.g. "assigns an integer to a string property, but does not cause a compilation error because the property is at a depth of 11"
